crates/linking/src/zksync.rs

```diff
@@ -0,0 +1,466 @@
+                // skip already linked
+                continue;
+            }
+
+            match Self::zk_link(
+                &contracts,
+                // we strip here _only_ so that the target matches what's in `contracts`
+                // but we want to return the full id in the `linked_artifacts`
+                &id.clone().with_stripped_file_prefixes(&self.linker.root),
+                libraries,
+                &self.compiler,
+            ) {
+                Ok(linked) => {
+                    *contracts.entry(id.clone()).or_default() = linked.clone();
+
+                    // persist linked contract for successive iterations
+                    linked_artifacts.push((id, CompactContractBytecode::from(linked)));
+                }
+                // contract was ignored, no need to add it to the list of linked contracts
+                Err(ZkLinkerError::MissingFactoryDeps(fdeps)) => {
+                    // attempt linking again if some factory dep remains unlinked
+                    // this is just in the case where a previously unlinked factory dep
+                    // is linked with the same run as `id` would be linked
+                    // and instead `id` remains unlinked
+                    // TODO(zk): might be unnecessary, observed when paths were wrong
+                    let mut ids = fdeps
+                        .iter()
+                        .flat_map(|fdep| {
+                            contracts.iter().find(|(id, _)| {
+                                // strip here to match against the fdep which is stripped
+                                let id =
+                                    (*id).clone().with_stripped_file_prefixes(&self.linker.root);
+                                id.source.as_path() == Path::new(fdep.filename.as_str()) &&
+                                    id.name == fdep.library
+                            })
+                        })
+                        // we want to keep the non-stripped
+                        .map(|(id, _)| id.clone())
+                        .peekable();
+
+                    // if we have no dep ids then we avoid
+                    // queueing our own id to avoid infinite loop
+                    // TODO(zk): find a better way to avoid issues later
+                    if let Some(sample_dep) = ids.peek() {
+                        // ensure that the sample_dep is in `contracts`
+                        if contracts.get(sample_dep).is_none() {
+                            return Err(ZkLinkerError::MissingFactoryDeps(fdeps));
+                        }
+
+                        targets.extend(ids); // queue factory deps for linking
+                        targets.push_back(id); // requeue original target
+                    }
+                }
+                Err(err) => return Err(err),
+            }
+        }
+
+        Ok(linked_artifacts.into_iter().collect())
+    }
+}
```

Comment:
Wonder if we can add some unit tests to validate DFS and other heuristics in this module.

Reply:
btw the DFS is from upstream, just slight edit to look thru factory deps too

Reply:
Ugh, it's rather annoying to do unit test here because we depend on compilation output, and we don't have a way (AFAIK) to make something out of thin air there.
I'd need to refactor the algorithms to be independent of `self` or mock it out